Parser/myreadline.c

```diff
@@ -24,25 +24,36 @@ static PyThread_type_lock _PyOS_ReadlineLock = NULL;
 int (*PyOS_InputHook)(void) = NULL;
 
 /* This function restarts a fgets() after an EINTR error occurred
-   except if PyOS_InterruptOccurred() returns true. */
+   except if _PyOS_InterruptOccurred() returns true. */
 
 static int
-my_fgets(char *buf, int len, FILE *fp)
+my_fgets(PyThreadState* tstate, char *buf, int len, FILE *fp)
 {
 #ifdef MS_WINDOWS
-    HANDLE hInterruptEvent;
+    HANDLE handle;
+    _Py_BEGIN_SUPPRESS_IPH
+    handle = (HANDLE)_get_osfhandle(fileno(fp));
+    _Py_END_SUPPRESS_IPH
+
+    /* bpo-40826: fgets(fp) does crash if fileno(fp) is closed */
+    if (handle == INVALID_HANDLE_VALUE) {
+        return -1; /* EOF */
+    }
```

Comment:
`_get_osfhandle` sets `errno` to `EBADF` if the fd is closed, so this is an error case that should return -2. That's consistent with Linux `fgets` for this case.

Reply:
my_fgets() is not a generic wrapper for fgets(), but a helper function for PyOS_StdioReadline() which doesn't use errno.

Reply:
If you see a bug in my change, please comment https://bugs.python.org/issue40826.

Reply:
I just thought it would be better to be consistent with the function's API -- returning -2 for an error -- and to be consistent how the same case would behave in Linux. Also, to not include a misleading `/* EOF */` comment for a case that has nothing to do with EOF.

Reply:
Well... the caller handles -1 and -2 the same way :-) This PR is a backport to 3.9. If you want to adjust the behavior, you can propose a PR on the master branch ;-)